src/vm/database/key_value_wrapper.rs

```diff
@@ -1,14 +1,71 @@
-use super::{MarfedKV, ClarityBackingStore};
+use super::{MarfedKV, ClarityBackingStore, ClarityDeserializable};
+use vm::Value;
 use vm::errors::{ InterpreterResult as Result };
 use chainstate::burn::BlockHeaderHash;
 use std::collections::{HashMap};
 use util::hash::{Sha512Trunc256Sum};
-use vm::types::QualifiedContractIdentifier;
+use vm::types::{QualifiedContractIdentifier, TypeSignature};
 use std::{cmp::Eq, hash::Hash, clone::Clone};
 
+#[cfg(rollback_value_check)]
+type RollbackValueCheck = String;
+#[cfg(not(rollback_value_check))]
+type RollbackValueCheck = ();
+
+#[cfg(not(rollback_value_check))]
+fn rollback_value_check(_value: &String, _check: &RollbackValueCheck) {}
+
+#[cfg(not(rollback_value_check))]
+fn rollback_edits_push<T>(edits: &mut Vec<(T, RollbackValueCheck)>, key: T, _value: &String) {
+    edits.push((key, ()));
+}
+// this function is used to check the lookup map when committing at the "bottom" of the
+//   wrapper -- i.e., when committing to the underlying store. for the _unchecked_ implementation
+//   this is used to get the edit _value_ out of the lookupmap, for used in the subsequent `put_all`
+//   command.
+#[cfg(not(rollback_value_check))]
+fn rollback_check_pre_bottom_commit<T>(edits: Vec<(T, RollbackValueCheck)>, lookup_map: &mut HashMap<T, Vec<String>>) -> Vec<(T, String)>
+where T: Eq + Hash + Clone {
+    for (_, edit_history) in lookup_map.iter_mut() {
+        edit_history.reverse();
+    }
+
+    let output = edits.into_iter().map(|(key, _)| {
+        let value = rollback_lookup_map(&key, &(), lookup_map);
+        (key, value)
+    }).collect();
+
+    assert!(lookup_map.len() == 0);
+    output
+}
+
+#[cfg(rollback_value_check)]
+fn rollback_value_check(value: &String, check: &RollbackValueCheck) {
+    assert_eq!(value, check)
+}
+#[cfg(rollback_value_check)]
+fn rollback_edits_push<T>(edits: &mut Vec<(T, RollbackValueCheck)>, key: T, value: &String)
+where T: Eq + Hash + Clone {
+    edits.push((key, value.clone()));
+}
+// this function is used to check the lookup map when committing at the "bottom" of the
+//   wrapper -- i.e., when committing to the underlying store.
+#[cfg(rollback_value_check)]
+fn rollback_check_pre_bottom_commit<T>(edits: Vec<(T, RollbackValueCheck)>, lookup_map: &mut HashMap<T, Vec<String>>) -> Vec<(T, String)>
```

Comment:
I'm having a hard time figuring out why these two methods exist.  Can you briefly explain what they both do, and why they're needed?

Reply:
There's essentially two implementations of the rollback wrapper -- one that tracks keys and values in the edit log _and_ materializes a hashmap for lookups (enabled via `rollback_value_check` flag), and one that materializes a hashmap for lookups and only tracks the edited keys in the edit log. The second reduces the overhead of the rollback wrapper by a factor of two. These two implementations need to many of the same things, but on commit and rollback, behave a little differently -- the first just does safety checks and the second needs to actually pull the values out of the hashmap to write the edit log out to the underlying storage.

Reply:
Do we need both implementations?  It looks like they do basically the same things?

Reply:
Yeah, they do pretty much the exact same thing, but the rollback_value_check has an additional panicking consistency check. This could be useful for correctness testing, but otherwise that version doesn't need to exist.

Reply:
Gotcha.  Thanks for clarifying.